### DISCUSSION TITLE
Support asynchronously resolved configuration

### DISCUSSION THREAD

**Discussion by pazams** [keystonejs/keystone#7273] [Feature requests]
`keystone.ts` interface is a default export of `KeystoneConfig`.
However, let's consider that some of the values in `KeystoneConfig` (e.g. db.url) needs to be fetched from a remote secret storage service. How should this type of async call work considering the interface for `keystone.ts` is a synchronous export of `KeystoneConfig`?

**Comment by pazams** [comment]
I created a reference solution here https://github.com/keystonejs/keystone/pull/7275

**Comment by dcousens** [comment]
Hi @pazams,
One of the problems with supporting a feature like this is that it breaks the nature of how we have treated configuration thus far;  and changes it into something that is _heavier_ than intended.

For example,  it might encourage asynchronous configuration practices that could be quite harmful during development when repeatedly restarting your Keystone process as part of the hot-reloading dev-tooling.

**Reply by pazams** [reply]
Thanks for your input @dcousens 

> heavier than intended

This solution proposed in https://github.com/keystonejs/keystone/pull/7275 is non-breaking to any keystone based project. Therefore if we can formally define what will be considered heavier, I can run some benchmarks and we can assess based on actual data if this feature qualifies as heavy?

> For example, it might encourage asynchronous configuration practices that could be quite harmful during development when repeatedly restarting your Keystone process as part of the hot-reloading dev-tooling.

I'd like to challenge this. I think this just increases the possibilities without necessarily pushing anyone to async. If thus far all known projects used configs based on values obtained from sync `process.env`, these configs are just probably going to remain unchanged.

On the other hand, projects with a product requirement of having secret values to be stored on 3rd parties (e.g. https://www.vaultproject.io/ or https://cloud.google.com/secret-manager ), will now also be able to use keystone as currently the synchronous config prevents that. It is only for these types of projects that devs will have the burden of keeping local dev-tooling snappy. But for these projects, having that burden is better than not being able to use keystone at all. Also I think it's a trivial solution for these projects to condition the key-fetching mode based on `NODE_ENV`, something along
```ts
type keyAccessor = (s: string) => Promise<string>

const envKeyAccessor: keyAccessor =  key => {
  const ret = process.env[key] ?? ""
  return Promise.resolve(ret)
}

const vaultKeyAccessor: keyAccessor =  key => {
  return callVaultService(key)
}

const getKey: keyAccessor =  key => {
  switch (process.NODE_ENV) {
	case 'production':
          return vaultKeyAccessor(key)
	  break
	default:
          return envKeyAccessor(key)
	  break
  }
}
```

**Reply by dcousens** [reply]
@pazams ~I would recommend~ (**edit**: this is complicated, and for reasons @pazams points out, environment variables might not be suitable for everyone) if you are constrained in this way, you can fetch the secrets using an asynchronous process and place them in the environment before running `keystone start`.

**edit 3**: If environment variables are not an option for you - because you want to adhere to the [OWASP](https://cheatsheetseries.owasp.org/cheatsheets/Secrets_Management_Cheat_Sheet.html#51-injection-of-secrets-file-in-memory) recommendations - [you could pass the configuration using `stdin` and read that synchronously.](https://stackoverflow.com/questions/37049439/is-passing-application-configuration-in-stdin-a-secure-alternative-to-environmen)

**Reply by pazams** [reply]
> @pazams I would recommend - that for users who are constrained in this way - to fetch the secrets using an asynchronous process and place them in the environment before running `keystone start`.

@dcousens some project and teams have different security requirments. Storing secrets in the environment as you suggest is not something some projects can consider. More generally, some might say "**Don’t store secrets in code, config files or pass them through environment variables.**" (Just an [OWASP](https://owasp.org/www-project-proactive-controls/v3/en/c8-protect-data-everywhere) recommendation... nothing major 😉 ).

I gave up on this project, so I won't be pushing for this anymore. Thanks for your consideration though.

**Reply by dcousens** [reply]
@pazams the [emphasis that OWASP is trying to make](https://cheatsheetseries.owasp.org/cheatsheets/Secrets_Management_Cheat_Sheet.html#51-injection-of-secrets-file-in-memory) is for the process (or child processes, or other processes) not to have access to secrets after startup; in the event a malicious actor can dump them later.
You might need to apply the same security considerations to whatever your asynchronous approach is too, if that matters to you.

That said, the benefits of external secrets management is auditing and rotation (and hopefully scoped access), and we want to help support that set of use cases, so this discussion is helpful as signal for having that support added :pray:

Let us know @pazams if you had any feedback as to if anything else prevented you from using Keystone 6 in your project :+1:

**Reply by pazams** [reply]
> and we want to help support that set of use cases

If you want to support it, then why reject a simple, non-breaking, [PR solution I provided](https://github.com/keystonejs/keystone/pull/7275) to solve it?  Sorry, you don't owe me any explainations, I just don't understand this type of communication.

> Let us know @pazams if you had any feedback as to if anything else prevented you from using Keystone 6 in your project 👍

No, just this. I actually enjoyed expirementing with keystone 6 and I recommended others to look into it. Any project though that is forced to use secrets management, this issue will be a non-starter in my opinion.

**Reply by dcousens** [reply]
@pazams the breadth and scope of the Keystone project and the features therein, that each need to be maintained, is wide.

The team attempts to maintain an exclusive focus on what we can support, in the short to medium term, rather than adding support for new features which may limit our velocity to introduce breaking changes with a low impact on the wider community, later.
Our initial team evaluation of #7275 was that the added limitation of requiring `await` for configuration introduced a barrier to future changes that aren't warranted by the community feedback and desire for this feature.

I understand that can be frustrating for end developers like yourself, especially when you have actively made the effort to make the pull request and contribute in a meaningful and positive way, but it's a balance that we have to maintain to keep the project stable.

If that position changes at any time, we'll re-evaluate your pull request, but as is, we didn't want to move forward with that feature for now.
